microshift_running_apps/microshift-operators-olm.adoc

```diff
@@ -0,0 +1,62 @@
+:_mod-docs-content-type: ASSEMBLY
+[id="olm-with-microshift"]
+= Using Operator Lifecycle Manager with {microshift-short}
+include::_attributes/attributes-microshift.adoc[]
+:context: microshift-operators-olm
+
+toc::[]
+
+OLM is used in {microshift-short} for installing and running optional link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html/architecture/control-plane#olm-operators_control-plane[add-on Operators].
+
+* Cluster Operators as applied in {ocp} are not used in {microshift-short}.
+* You can add your own catalogs for the Operators you want to use with your applications.
+* You must use the CLI to conduct OLM activities with {microshift-short}. The console and OperatorHub GUIs are not available, but you can take all the same actions with the CLI.
+* Install the link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html/cli_tools/opm-cli#cli-opm-install[`opm` CLI]
+
+[IMPORTANT]
+====
+Before using an Operator, verify with the provider whether the Operator is supported on {product-title}.
+====
+
+[id="microshift-file-based-olm-catalogs"]
+== File-based Operator catalogs
+
+To run {microshift-short} with your own curated catalogs and OLM, use the file-based catalog structure. For a complete explanation of the file-based structure, see the "Additional resources" section.
+
+include::modules/olm-fb-catalogs-example.adoc[leveloffset=+2]
+include::modules/olm-fb-catalogs-guidelines.adoc[leveloffset=+2]
+include::modules/olm-creating-fb-catalog-image.adoc[leveloffset=+2]
+
+[role="_additional-resources"]
+.Additional resources
+* link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.15/html/cli_tools/opm-cli[`opm` CLI reference]
+* link:https://https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html/operators/understanding-operators#olm-file-based-catalogs_olm-packaging-format[File-based catalogs]
+
+include::modules/olm-creating-catalog-from-index.adoc[leveloffset=+2]
+
+[id="microshift-using-olm-disconnected"]
+== Using OLM in disconnected clusters
```

Comment:
@ShaunaDiaz for https://70128--ocpdocs-pr.netlify.app/microshift/latest/microshift_running_apps/microshift-operators-olm#microshift-using-olm-disconnected,
I am confused on what this section does.
it is to install OLM component in restricted network, or install OLM-based operator in restricted network.

if it is to install OLM component in restricted network,
A) for See [Mirroring container images for disconnected installations](https://70128--ocpdocs-pr.netlify.app/microshift/latest/microshift_install/microshift-deploy-with-mirror-registry#microshift-deploy-with-mirror-registry), the link does not handle olm component payload because it is not in microshift-release-info*.noarch.rpm. so need to change the content of the link
B) no need 
```console
Use a workstation with full internet access to create and push local mirrors of the catalog content to a mirror registry.

Add a CatalogSource to the cluster so that the OLM catalog Operator can use the local catalog on the mirror registry.
```

if it is to install OLM-based operator in restricted network,
A) I do not think [Mirroring container images for disconnected installations](https://70128--ocpdocs-pr.netlify.app/microshift/latest/microshift_install/microshift-deploy-with-mirror-registry#microshift-deploy-with-mirror-registry) is OK to mirror index image and bundle image of the operator.  please contact with OLM expert to correct it.
B), for See [Using Operator Lifecycle Manager on restricted networks](https://access.redhat.com/documentation/en-us/openshift_container_platform/4.14/html/operators/administrator-tasks#olm-restricted-networks) for more information,  the content of the link is for ocp and not for microshift. it means some steps of the link can not be done for micorshift, for example, no OperatorHub, no UI, no ISCP, no IDMS, no ITMS.

so, I do not think referencing ocp doc link directly is not good. it will make customer confused.

@kevinrizza  please correct me if I am wrong. thanks

Reply:
This section of the docs is definitely referring to adding olm based operators to a disconnected cluster. I don't have a strong opinion about whether we need an entirely separate copy of the documentation, but the procedure to actually do the mirroring is definitely what we point folks to regardless of whether or not they'll have access to the openshift console. But yes, we should probably firm up our information around how to mirror images without ICSPs if we're not expecting users who already interact with microshift to know how to modify the container runtime bits directly.

Reply:
@pmtk Do we need an OLM engineer to consult with us to "...firm up our information around how to mirror images without ICSPs if we're not expecting users who already interact with microshift to know how to modify the container runtime bits directly." I added a link to the disconnected install in the bullet about microshift itself. Is that helpful?

Reply:
There is a tool that exists explicitly to solve this problem for OLM operators. You can use `oc-mirror` to generate the list of operator bundle and related images from the catalog. The bit that users will have to do manually is take that image list and manually configure cri-o once oc-mirror does the actual mirroring process

Reply:
when we install olm-based operator in restricted cluster, we have two steps to prepare it before creating catsrc and installing operator.
1, mirror images related: index image, bundle images, operator images etc
2, create iscp

for 2, it is mentioned not to use ISCP and config crio directly.
for 1, we also can use `oc adm mirror` here is reference https://docs.openshift.com/container-platform/4.14/installing/disconnected_install/installing-mirroring-installation-images.html#installation-mirror-repository_installing-mirroring-installation-images, for example
```console
oc adm catalog mirror quay.io/olmqe/nginx-ok-index:v1399-fbc-multi-digest-new dhcp-1-235-13.arm.eng.rdu2.redhat.com:5000 --manifests-only
# you will get catsrc yaml, iscp yaml, mapping.txt. ignore icsp yaml
oc image mirror -f manifests-nginx-ok-index-1702537156/mapping.txt --filter-by-os=.*

```
and then the related images are mirror into internal registry.

after step1 and step2, can create catsrc and install operator.

Reply:
> know how to modify the container runtime bits directly

Which part is this? You mean configuring cri-o to use mirror or something else?

> You can use oc-mirror to generate the list of operator bundle and related images from the catalog

We expect users to create their own catalogs, not mirror openshift catalog - especially there's no guarantee that operators from openshift catalog would work correctly on microshift.

Reply:
I'm not sure what is topic of this thread anymore.

To use MicroShift & OLM in disconnected envs we need to:
- Mirror microshift's images using json from microshift-release-info RPM
- Mirror OLM images using json from microshift-olm (in 4.15.1 we want to add microshift-olm-release-info) - @ShaunaDiaz I guess we need to add this because current procedure only mentions microshift-release-info.
- Mirror/store images of the CatalogSource and operator user want to use
- Create CatalogSource that will use original image refs if they're mirrored OR use internal mirror refs

Assuming users will craft their own catalogs and not reuse openshift's catalog/marketplace, what else are we missing? I think they could just use already existing procedure to mirror more images by supplying more image refs.

Reply:
@pmtk I just don't think "create their own catalogs from scratch" is a reasonable or feasible expectation from our users. That would require not only the bundle references themselves, but the prior backdated knowledge of the order that the operators were inserted because today `opm` mutates the update graph on each individual insert. Yes, it's definitely true that everything can't be mirrored as is, but it is going to be far far simpler for customers to prune the official catalogs to contain just the operators that can run on microshift and they want to install and then mirror that pruned catalog.

Reply:
> @pmtk I just don't think "create their own catalogs from scratch" is a reasonable or feasible expectation from our users. 

The feature we're providing is custom catalogs for OEM partners to build to bundle with MicroShift when they deliver it to their end users. That catalog may contain Red Hat supported operators, but may also contain other content from that partner.

If we need to have a quick call to set expectations here, let's organize one.

Reply:
yep, happy to jump on whenever.

Reply:
I think the confusion is from the fact that the tooling that is used for offline/airgapped installations is ALSO used to created custom catalog images. I will organize a call to discuss.

Reply:
@pmtk I added the olm references to the "getting container references" module; remove the OCP link as well, but we probably want some procedure for users.

Reply:
for `dhcp-1-235-13.arm.eng.rdu2.redhat.com:5000` private registry, it is just one example. 
it could be one same private registry for both micorshift and operator.